pde and unit test inconsistencies

stonecoldhughes opened the pull request:
fokkerplanck_2 passes time_advance_tests in C++. Several other PDEs are either inconsistent or the gold data generation parameters in Matlab differ from the C++ ones used in tests.

closes #270 .

bmcdanie reviewed (changes requested):
please trim all the unnecessary/unrelated files from this PR @stonecoldhughes . there are hundreds of cmake files etc. that should not have been committed, and it's burying the diffs for the files I need to review.

Commit:
inconsistency between Matlab and C++ PDE implementations and unit tests

fixed unit test inconsistencies

Modified legendre weights routine to use floating-point interval bounds

Updated fokkerplanck2_complete and a few others to match Matlab

Added RAII-friendly test function for unit tests in time_advance_tests.cpp

bmcdanie commented on src/time_advance_tests.cpp:
```diff
@@ -24,192 +24,97 @@ static distribution_test_init const distrib_test_info;
-void time_advance_test(int const level, int const degree, PDE<P> &pde,
-                       int const num_steps, std::string const filepath,
-                       bool const full_grid = false, P const tol_factor = 1e-10)
+void time_advance_test(parser const &parse, std::string const &filepath,
```
why did you change this function?

stonecoldhughes replied:
All the arguments to the original version of the function are now encapsulated in the parser class, so it is more concise this way. With this configuration, the body of the unit tests are only responsible for defining state via a parser object and then this routine builds the PDE from it. Previously, there were 2 separate test functions for time_advance - for explicit and implicit. Now the unified version infers which should be called from the parser object.

bmcdanie replied:
ok, that's an improvement.

bmcdanie commented on src/time_advance_tests.cpp:
```diff
@@ -223,112 +128,145 @@ void implicit_time_advance_test(int const level, int const degree, PDE<P> &pde,
-                                  "diffusion1/diffusion1_e_sg_l4_d4_t";
+                                  "diffusion1_sg_l4_d4_t";
+    TestType const tol_factor =
+        std::is_same<TestType, double>::value ? 1e-12 : 1e-2;
```
is this one looser than it used to be?

stonecoldhughes replied:
Yes - after updating the pde to match the reference implementation, the tolerances had to be loosened.

bmcdanie replied:
ok - did we see any of that in other component tests for this pde? we're loosening here, are we loosening for this PDE in coefficient generation, source vectors, boundary conditions?

bmcdanie replied:
checked - the answer is no.

bmcdanie commented on .gitignore:
```diff
@@ -18,3 +18,4 @@ contrib/lperftools
 contrib/valgrind
 contrib/hdf5
 contrib/highfive
+contrib/kronmult
```
thanks

bmcdanie commented on src/pde/pde_diffusion1.hpp:
```diff
@@ -21,16 +21,18 @@ class PDE_diffusion_1d : public PDE<P>
   initial_condition_dim0(fk::vector<P> const &x, P const t = 0)
   {
-    static double p     = -2.0 * PI * PI;
+    static double p     = -2.0 * nu * nu;
```
could be constified.

stonecoldhughes replied:
done

bmcdanie commented on src/pde/pde_diffusion1.hpp:
```diff
@@ -93,21 +87,21 @@ class PDE_diffusion_1d : public PDE<P>
   static P source_0_t(P const t)
   {
-    static double neg_two_pi_squared = -2.0 * PI * PI;
+    static double coefficient = -2.0 * nu * nu;
```
const

stonecoldhughes replied:
done

bmcdanie commented on src/pde/pde_fokkerplanck2_complete.hpp:
```diff
@@ -73,16 +74,16 @@ class PDE_fokkerplanck_2d_complete : public PDE<P>
   static auto constexpr psi = [](P x) {
     auto dphi_dx = 2.0 / std::sqrt(M_PI) * std::exp(-std::pow(x, 2));
```
constify

stonecoldhughes replied:
done

bmcdanie commented on src/pde/pde_fokkerplanck2_complete.hpp:
```diff
@@ -114,13 +115,51 @@ class PDE_fokkerplanck_2d_complete : public PDE<P>
+    P h = 20.0 / N;
+    P Q = 0;
+
+    auto function = [](fk::vector<P> const &p) -> fk::vector<P> {
```
const

stonecoldhughes replied:
done

bmcdanie commented on src/time_advance_tests.cpp:
```diff
@@ -499,151 +480,182 @@ TEMPLATE_TEST_CASE("time advance - fokkerplanck_1d_4p1a", "[time_advance]",
+    int const level  = 5;
+    int const degree = 3;
+
+    TestType const tol_factor =
```
good

bmcdanie commented on src/time_advance_tests.cpp:
```diff
@@ -499,151 +480,182 @@ TEMPLATE_TEST_CASE("time advance - fokkerplanck_1d_4p1a", "[time_advance]",
 TEMPLATE_TEST_CASE("time advance - fokkerplanck_2d_complete", "[time_advance]",
                    float, double)
 {
   /* FIXME - these tolerances are way too high. Different parameters are likely
```
take this comment out, it's fixed

stonecoldhughes replied:
done

bmcdanie reviewed (changes requested):
looks good overall. let's tighten up the const in a couple of areas (these were not const prior but we can fix now). also, let's discuss briefly why the test function change. not opposed, just want to know why we switched.

bmcdanie commented on src/pde/pde_diffusion1.hpp:
```diff
@@ -21,16 +21,18 @@ class PDE_diffusion_1d : public PDE<P>
   {
-    static double p     = -2.0 * PI * PI;
-    P const coefficient = std::exp(p * t);
+    static const double p = -2.0 * nu * nu;
```
west const! you wound me

stonecoldhughes replied:
lol fixed
